pkg/workloadlearning/handle.go

```diff
@@ -217,4 +303,408 @@ func (handle *Handle) SaveTableReadCostMetrics(metrics map[ast.CIStr]*TableReadC
 	if err != nil {
 		logutil.BgLogger().Warn("commit txn failed when saving table cost metrics", zap.Error(err))
 	}
+	// TODO: saving the workload job info such as start end time into workload_jobs table
+}
+
+func checkCrossDB(sql string) (bool, error) {
+	// TODO check duplicate alias name
+	p := parser.New()
+	// notes: this is only use for check whether query cross db, so the charset and collation doesn't matter in here.
+	stmtNodes, _, err := p.Parse(sql, "", "")
+	if err != nil {
+		return false, err
+	}
+
+	extractor := &TableNameExtractor{}
+	if len(stmtNodes) != 1 {
+		// TODO: support multi-statement in one single statement_record
+		return false, nil
+	}
+	for _, stmt := range stmtNodes {
+		stmt.Accept(extractor)
+		if len(extractor.DBs) > 1 {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// extractScanAndMemoryFromBinaryPlan abstract the scan time and memory usage from one plan string
+// The input is the field BINARY_PLAN in **one row** of table mysql.CLUSTER_TIDB_STATEMENTS_STATS
+// The result should be **list** because every query maybe related to multiple tables such as: select * from t1, t2
+// Every scan operator in the plan string should be extracted to one TableReadCostMetrics without any grouping
+func extractScanAndMemoryFromBinaryPlan(binaryPlan string) ([]*TableReadCostMetrics, error) {
+	// Step1: convert binaryPlan to tipb.ExplainData
+	protoBytes, err := plancodec.Decompress(binaryPlan)
+	if err != nil {
+		return nil, err
+	}
+	explainData := &tipb.ExplainData{}
+	err = explainData.Unmarshal(protoBytes)
+	if err != nil {
+		return nil, err
+	}
+	if explainData.DiscardedDueToTooLong {
+		return nil, fmt.Errorf("plan is too long, discarded it, failed to convert to ExplainData")
+	}
+	if !explainData.WithRuntimeStats {
+		return nil, fmt.Errorf("plan is not with runtime stats, failed to convert to ExplainData")
+	}
+
+	// Step2: abstract the scan time and memory usage from explainData
+	operatorExtractMetrics := make([]*TableReadCostMetrics, 1)
+	// extract scan and memory from main part plan
+	err = extractMetricsFromOperatorTree(explainData.Main, operatorExtractMetrics)
```

Comment:
The slice `operatorExtractMetrics` is not going to be updated if we use it in this way.
<img width="549" alt="image" src="https://github.com/user-attachments/assets/b58d8030-25a6-4734-a1b8-215f450fc06f" />

A common way to solve this is like below, just let the function return the newly appended slice:
```
operatorExtractMetrics, err = extractMetricsFromOperatorTree(explainData.Main, operatorExtractMetrics)
```